torchmetrics/functional/image/sam.py

```diff
@@ -0,0 +1,115 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Tuple, Union
+
+import torch
+from torch import Tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.regression.cosine_similarity import cosine_similarity
+from torchmetrics.utilities.checks import _check_same_shape
+from torchmetrics.utilities.distributed import reduce
+
+
+def _sam_update(preds: Tensor, target: Tensor) -> Tuple[Tensor, Tensor]:
+    """Updates and returns variables required to compute Spectral Angle Mapper. Checks for same shape and type of
+    the input tensors.
+
+    Args:
+        preds: Predicted tensor
+        target: Ground truth tensor
+    """
+
+    if preds.dtype != target.dtype:
+        raise TypeError(
+            "Expected `preds` and `target` to have the same data type."
+            f" Got preds: {preds.dtype} and target: {target.dtype}."
+        )
+    _check_same_shape(preds, target)
+    if len(preds.shape) != 4:
+        raise ValueError(
+            "Expected `preds` and `target` to have BxCxHxW shape."
+            f" Got preds: {preds.shape} and target: {target.shape}."
+        )
+    return preds, target
+
+
+def _sam_compute(
+    preds: Tensor,
+    target: Tensor,
+    reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
+) -> Union[Tensor, Tuple[Tensor, Tensor]]:
+    """Computes Spectral Angle Mapper.
+
+    Args:
+        preds: estimated image
+        target: ground truth image
+        reduction: a method to reduce metric score over labels.
+
+            - ``'elementwise_mean'``: takes the mean (default)
+            - ``'sum'``: takes the sum
+            - ``'none'``: no reduction will be applied
+
+    Example:
+        >>> preds = torch.rand([16, 1, 16, 16], generator=torch.manual_seed(42))
+        >>> target = torch.rand([16, 1, 16, 16], generator=torch.manual_seed(123))
+        >>> preds, target = _sam_update(preds, target)
+        >>> _sam_compute(preds, target)
+        tensor(0.7377)
+    """
+    B, C, H, W = preds.shape
+    preds = preds.reshape(B, C, H * W)
+    target = target.reshape(B, C, H * W)
```

Comment:
That reshaping leads to the computation of the **spatial** angle, whereas the **spectral** angle should be computed. The spectral dimension C should be put as last dim, as said in [this comment](https://github.com/PyTorchLightning/metrics/issues/799#issuecomment-1063897771).
The error can be spotted because computing the sam with `reduction="none"` will lead to a shape (B, C) whereas it should be (B, H, W), as the angle should be computed between each pixel, in the spectral dimension.
The test on shape could be added to the test suite.

```suggestion
    preds = torch.movedim(preds, 1, -1)
    target = torch.movedim(target, 1, -1)
```

Reply:
Thank you for your comment. This is my mistake to compute the spectral angle. In addition, I will add check the dimension of preds and target to make sure that they should be larger than 1 to compute the cosine similarity of spectral angle.